lib/vdsm/storage/merge.py

```diff
@@ -207,18 +205,29 @@ def _extend_base_allocation(base_vol, top_vol):
     if not (base_vol.is_block() and base_vol.getFormat() == sc.COW_FORMAT):
         return
 
-    base_alloc = base_vol.getVolumeSize()
-    top_alloc = top_vol.getVolumeSize()
-    vol_chunk_size = config.getint('irs', 'volume_utilization_chunk_mb') * MiB
-    potential_alloc = base_alloc + top_alloc + vol_chunk_size
-    # TODO: add chunk_size only if top is leaf.
-    capacity = base_vol.getCapacity()
-    max_alloc = utils.round(capacity * sc.COW_OVERHEAD, MiB)
-    actual_alloc = min(potential_alloc, max_alloc)
-    actual_alloc = utils.round(actual_alloc, MiB)
-    actual_alloc_mb = actual_alloc // MiB
+    # Measure the subchain from top to base. This gives us the required
+    # allocation for merging top into base.
+    log.debug("Measuring sub chain top=%r base=%r",
+              top_vol.volUUID, base_vol.volUUID)
+    measure = qemuimg.measure(
+        top_vol.getVolumePath(),
+        format=qemuimg.FORMAT.QCOW2,
+        output_format=qemuimg.FORMAT.QCOW2,
+        is_block=True,
+        base=base_vol.getVolumePath())
+    log.debug("Measure result: %s", measure)
+
+    # When merging we always copy the bitmaps from the top to base. Measure
+    # gives us the size of the bitmaps in top *and* base, so this may allocate
+    # more than needed, but bitmaps are small so it should be good enough.
+    required_size = measure["required"] + measure.get("bitmaps", 0)
+
+    optimal_size = base_vol.optimal_cow_size(
+        required_size, base_vol.getCapacity(), base_vol.isLeaf())
```

Comment:
Testing reveal that this is wrong, should be top_vol.isLeaf().

We need to improve the test to test both internal and active merge.